### ISSUE TITLE
Get host port number from DockerInspectContainer

### ISSUE THREAD

**Issue by mgurov** [bmuschko/gradle-docker-plugin#342]
I'd like to achieve #316 , but cannot make the host ports mapping to appear in the DockerInspectContainer response.

I have the following command to create container: 

````groovy
task createContainer(type: DockerCreateContainer) {
    ....
    exposePorts("tcp", [5555, 5556])
    portBindings = ['5555', ':5556']
}
````

I see the ports to be mapped via `docker ps`: `0.0.0.0:32828->5555/tcp, 0.0.0.0:32827->5556/tcp`. Now I'd like to use these `32828` and `32827` to run the tests against the started containers, but cannot find a good way to do that. 

I tried `DockerInspectContainer`, and the response seems to be legit except for `getNetworkSettings().getPorts()` is always `null`. 

Am I using a wrong method to get the mapped ports or is there a bug to fix somewhere? 

com.bmuschko:gradle-docker-plugin:3.0.5

````
$ ./gradlew --version

------------------------------------------------------------
Gradle 3.1
------------------------------------------------------------

Build time:   2016-09-19 10:53:53 UTC
Revision:     13f38ba699afd86d7cdc4ed8fd7dd3960c0b1f97

Groovy:       2.4.7
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_92 (Oracle Corporation 25.92-b14)
OS:           Mac OS X 10.11.6 x86_64

$ docker version
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Wed Jan 18 16:20:26 2017
 OS/Arch:      darwin/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Wed Jan 18 16:20:26 2017
 OS/Arch:      linux/amd64
 Experimental: true
````

**Comment by cdancy** [comment]
You can use portBindings like so:

```
    portBindings = ["11222:5432"]
```

Where port 11222 on the host will be mapped to port 5432 inside the container.

**Comment by mgurov** [comment]
@cdancy thank you for your feedback, but I guess I wasn't very clear. I *do* use `it.networkSettings.ports`, but it is always returns `null` for me despite what `docker ps` prints. 

I cannot use `portBindings = ["11222:5432"]` in generic case because we would like to be able to run tests in parallel without clashes for ports on our CI environment. 

What I could do is to find a free port and assign it for the port binding, but that gets a little bit involved as I need to make sure that a parallel build doesn't claim the same port.

I did hope that the `gradle-docker-plugin` would allow us to leverage already existing docker's function which is so nicely provided by maven [docker plugin](https://dmp.fabric8.io/#start-port-mapping) (see the second form where the plugin itself binds the newly allocated port to a variable). 

Shall I create a more specific issue about `inspectResponse.networkSettings.ports` been null?

**Comment by cdancy** [comment]
@mgurov if you want a random port just leave out everything to the left of the colon:

```
portBindings = [":5432"]
```

**Comment by mgurov** [comment]
@cdancy I did that in my example: `portBindings = ['5555', ':5556']`. The problem is not instructing docker, it is to get the allocated host port back into the gradle build.

**Comment by cdancy** [comment]
@mgurov can you wire the below into your code and paste the output here:

```
task inspect(type: DockerInspectContainer) {
    onNext {
        def networkSettings = it.networkSettings
        println networkSettings // or do whatever else you'd like with these properties
    }
}
```

**Comment by mgurov** [comment]
@cdancy 

````
com.github.dockerjava.api.model.NetworkSettings@bb85bb1[bridge=,sandboxId=,hairpinMode=false,linkLocalIPv6Address=,linkLocalIPv6PrefixLen=0,ports=<null>,sandboxKey=,secondaryIPAddresses=<null>,secondaryIPv6Addresses=<null>,endpointID=,gateway=,portMapping=<null>,globalIPv6Address=,globalIPv6PrefixLen=0,ipAddress=,ipPrefixLen=0,ipV6Gateway=,macAddress=,networks={bridge=com.github.dockerjava.api.model.ContainerNetwork@724f0110[ipamConfig=<null>,links=<null>,aliases=<null>,networkID=,endpointId=,gateway=,ipAddress=,ipPrefixLen=0,ipV6Gateway=,globalIPv6Address=,globalIPv6PrefixLen=0,macAddress=]}]
````

**Comment by cdancy** [comment]
@mgurov what version of docker are you using? And can you confirm that the container is actually running and not just created?

**Comment by mgurov** [comment]
@cdancy docker 13.0 on mac os x. Container is running, I see logs and can connect to the ports manually. `docker inspect` shows following network settings: 

````json
"Ports": {
                "5555/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "32840"
                    }
                ],
                "5556/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "32839"
                    }
                ]
            },
````

**Comment by cdancy** [comment]
@mgurov hmm ... same setup here. Can you post a reproducible snippet so that I can give it a go here?

**Comment by mgurov** [comment]
@cdancy will try to compose something public, probably not earlier than the weekend though

**Comment by mgurov** [comment]
@cdancy my apologies. The container wasn't indeed started, only created. Wrong dependency for the inspect task. 

Would documenting a way to get a random port be a good idea? Let me know if I could contribute this way. In my experience I see that quite some people who integrate docker into their build are hitting this issue and maybe extending the "functional test" example would make things bit easier from the start.

**Comment by cdancy** [comment]
@mgurov if you're up for it send in a PR and maybe add another example, or add to an existing example, on our `README.md` file which demonstrates the use of a random port.

**Comment by dcoraboeuf** [comment]
Hi,

Pity this issue is closed. Would be very nice to have some documentation about the way to get a random port from a container started by another task. Very useful for functional tests.

@mgurov , did you get this working in the end?

Thanks,
Damien.

**Comment by cdancy** [comment]
@dcoraboeuf we have examples in our code under the fucntionalTests. This should get you started:

```
task inspect(type: DockerInspectContainer) {
    onNext {
        it.networkSettings.getPorts().getBindings().each { k, v ->
            def key = "" + k.getPort()
            def value = "" + v[0].hostPortSpec
            println "Port ${key} is exposed on host as ${value}"
        }
    }
}
```

**Comment by dcoraboeuf** [comment]
Thanks @cdancy . I'll implement this.

**Comment by mgurov** [comment]
@dcoraboeuf my approach for the reference. 
Updates system property `defaults.database.url` according to the random port allocated for postgres database container. Additionally allows predefining the port via `pg.port` system property. 

Warning: non-idiomatic gradle. 

```groovy
def onPostgresPortResolved = { p ->
    tasks.withType(Test) {
        def jdbcUrl = "jdbc:postgresql://localhost:${p.getHostPortSpec()}/database_name"
        systemProperty "defaults.database.url", jdbcUrl
    }
}
def dockerPortsMap = [5432: ["key": "pg.port", "onResolved": onPostgresPortResolved]]

task createPgContainer(type: DockerCreateContainer) {
    targetImageId { 'postgres:9.5' }
    portBindings = dockerPortsMap.collect { port, entry -> "${System.getProperty(entry.key, '')}:${port}" }
}

task startPgContainer(type: DockerStartContainer) {
    dependsOn createPgContainer
    targetContainerId { createPgContainer.getContainerId() }
}

task stopPgContainer(type: DockerStopContainer) {
    targetContainerId { createPgContainer.getContainerId() }
}

task inspectPgPorts(type: DockerInspectContainer) {
    dependsOn startPgContainer
    targetContainerId { createPgContainer.getContainerId() }
    onNext {
        it.networkSettings.ports.getBindings().each { exposedPort, bindingArr ->
            def portMapping = dockerPortsMap[exposedPort.getPort()]
            def consumer
            if (portMapping.onResolved) {
                consumer = portMapping.onResolved
            } else {
                consumer = {
                    println("${portMapping.key} -> ${it.getHostPortSpec()}, no action taken on this docker mapping")
                }
            }
            bindingArr.each consumer
        }
    }
}

test.dependsOn inspectPgPorts
test.finalizedBy stopPgContainer
```

**Comment by cdancy** [comment]
@orzeh really love seeing people take advantage of the reactive functionality.

**Comment by cdancy** [comment]
@dcoraboeuf feel free to send in a PR with an example or updated documentation or both!

**Comment by dcoraboeuf** [comment]
(y) I do this now (extract of my `build.gradle` file):

```groovy
task createAppContainer(type: DockerCreateContainer) {
    dependsOn buildDockerImage
    targetImageId { buildDockerImage.getImageId() }
    portBindings = [":8080"]
}
task startAppContainer(type: DockerStartContainer) {
    dependsOn createAppContainer
    targetContainerId { createAppContainer.getContainerId() }
}
task portsOfAppContainer(type: DockerInspectContainer) {
    dependsOn startAppContainer
    targetContainerId { createAppContainer.getContainerId() }
    onNext {
        it.networkSettings.ports.getBindings().each { exposedPort, bindingArr ->
            if (exposedPort.port == 8080) {
                int portBinding = bindingArr.first().hostPortSpec as int
                String url = "http://${testDockerHost}:${portBinding}"
                tasks.extensionTest {
                    systemProperty 'acceptance.url', url
                }
            }
        }
    }
}
task extensionTest(type: Test) {
    dependsOn portsOfAppContainer
    finalizedBy removeAppContainer
    include '**/*Acceptance.class'
}
```

Thanks for the initial idea.
